tensorboard/webapp/widgets/histogram/histogram_v2_component.ts

```diff
@@ -55,83 +77,178 @@ export class HistogramV2Component implements OnChanges {
+    const countMax = getMax(data, (datum) => {
+      return getMax(datum.bins, ({y}) => y);
+    });
 
-  private updateYScaleAndDomain(data: HistogramData) {
-    this.yScale =
+    // === Create scale and set the domain. ===
+    const binScale = d3.scaleLinear().domain([binMin, binMax]).nice();
+    const temporalScale =
       this.mode !== HistogramMode.OVERLAY &&
       this.timeProperty == TimeProperty.WALL_TIME
         ? d3.scaleTime()
         : d3.scaleLinear();
 
-    let domain: [number, number] = [0, 1];
+    const timeValues = data.map((datum) => this.getTimeValue(datum));
+    const {min: timeMin, max: timeMax} = getMinMax(timeValues, (val) => val);
+    const temporalDomain = [timeMin, timeMax];
+
+    temporalScale.domain(temporalDomain);
+    const countScale = d3.scaleLinear();
+    countScale.domain([0, countMax]);
+    const d3Color = d3.hcl(
+      this.colorScale ? this.colorScale(this.name) : '#000'
+    );
+    const d3ColorScale = d3.scaleLinear<HCLColor, string>();
+    d3ColorScale.domain(temporalDomain);
+
+    // === Set range on scales. ===
+    // x-axis or bin scale does not change depending on a mode.
+    binScale.range([0, width]);
+    d3ColorScale.range([d3Color.brighter(), d3Color.darker()]);
+    d3ColorScale.interpolate(d3.interpolateHcl);
+
+    // Explanation of the coordinate systems:
+    // When in the offset mode, we render in 2.5D. Y-axis both have temporal
+    // element while some space is allocated to show magnitude of counts. To
+    // make the coordinate system easier for the offset, we use `<g transform>`
+    // to locate the histogram at the correct temporal axis and let `countScale`
+    // only deal with the height of the histogram.
+    // When in overlay mode, we have very simple 2D where temporal axis is not
+    // used meaningfully and `countScale` act as the y-axis and thus spans
+    // `[height, 0]`.
     if (this.mode === HistogramMode.OVERLAY) {
-      const countMax = getMax(data, (datum) => {
-        return getMax(datum.bins, ({y}) => y);
-      });
-      domain = [0, countMax];
+      temporalScale.range([height, height]);
+      countScale.range([height, 0]);
     } else {
-      const yValues = data.map((datum) => {
-        switch (this.timeProperty) {
-          case TimeProperty.WALL_TIME:
-            return datum.wallTime;
-          case TimeProperty.STEP:
-            return datum.step;
-          case TimeProperty.RELATIVE:
-            return datum.wallTime - data[0].wallTime;
-        }
-      });
-      const {min: yMin, max: yMax} = getMinMax(yValues, (val) => val);
-      domain = [yMin, yMax];
+      const offsetAxisHeight =
+        this.mode === HistogramMode.OFFSET
+          ? height - layout.histogramHeight
+          : 0;
+      temporalScale.range([height - offsetAxisHeight, height]);
+      countScale.range([0, -layout.histogramHeight]);
     }
-    this.yScale.domain(domain);
+
+    return {binScale, d3ColorScale, countScale, temporalScale};
```

Comment:
Since `updateScales` does not actually update the property, consider either renaming to `computeScales` or making the method actually update the scales.